client/iOS/FreeRDP/ios_cliprdr.m

```diff
@@ -0,0 +1,524 @@
+
+	afc = (mfContext *)cliprdr->custom;
+
+	if (!afc)
+		return ERROR_INVALID_PARAMETER;
+
+	if ((rc = ios_cliprdr_send_client_capabilities(cliprdr)) != CHANNEL_RC_OK)
+		return rc;
+
+	if ((rc = ios_cliprdr_send_client_format_list(cliprdr)) != CHANNEL_RC_OK)
+		return rc;
+
+	afc->clipboardSync = TRUE;
+	return CHANNEL_RC_OK;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ios_cliprdr_server_capabilities(CliprdrClientContext *cliprdr,
+                                            const CLIPRDR_CAPABILITIES *capabilities)
+{
+	UINT32 index;
+	CLIPRDR_CAPABILITY_SET *capabilitySet;
+	mfContext *afc;
+
+	if (!cliprdr || !capabilities)
+		return ERROR_INVALID_PARAMETER;
+
+	afc = (mfContext *)cliprdr->custom;
+
+	if (!afc)
+		return ERROR_INVALID_PARAMETER;
+
+	for (index = 0; index < capabilities->cCapabilitiesSets; index++)
+	{
+		capabilitySet = &(capabilities->capabilitySets[index]);
+
+		if ((capabilitySet->capabilitySetType == CB_CAPSTYPE_GENERAL) &&
+		    (capabilitySet->capabilitySetLength >= CB_CAPSTYPE_GENERAL_LEN))
+		{
+			CLIPRDR_GENERAL_CAPABILITY_SET *generalCapabilitySet =
+			    (CLIPRDR_GENERAL_CAPABILITY_SET *)capabilitySet;
+			afc->clipboardCapabilities = generalCapabilitySet->generalFlags;
+			break;
+		}
+	}
+
+	return CHANNEL_RC_OK;
+}
+
+/**
+ * Function description
+ *
+ * @return 0 on success, otherwise a Win32 error code
+ */
+static UINT ios_cliprdr_server_format_list(CliprdrClientContext *cliprdr,
+                                           const CLIPRDR_FORMAT_LIST *formatList)
+{
+	UINT rc;
+	UINT32 index;
+	CLIPRDR_FORMAT *format;
+	mfContext *afc;
+
+	if (!cliprdr || !formatList)
+		return ERROR_INVALID_PARAMETER;
+
+	afc = (mfContext *)cliprdr->custom;
+
+	if (!afc)
+		return ERROR_INVALID_PARAMETER;
+
+	if (afc->serverFormats)
+	{
+		for (index = 0; index < afc->numServerFormats; index++)
+			free(afc->serverFormats[index].formatName);
+
+		free(afc->serverFormats);
+		afc->serverFormats = NULL;
+		afc->numServerFormats = 0;
+	}
+
+	if (formatList->numFormats < 1)
+		return CHANNEL_RC_OK;
+
+	afc->numServerFormats = formatList->numFormats;
+	afc->serverFormats = (CLIPRDR_FORMAT *)calloc(afc->numServerFormats, sizeof(CLIPRDR_FORMAT));
+
+	if (!afc->serverFormats)
+		return CHANNEL_RC_NO_MEMORY;
+
+	for (index = 0; index < afc->numServerFormats; index++)
+	{
+		afc->serverFormats[index].formatId = formatList->formats[index].formatId;
+		afc->serverFormats[index].formatName = NULL;
+
+		if (formatList->formats[index].formatName)
+		{
+			afc->serverFormats[index].formatName = _strdup(formatList->formats[index].formatName);
+
+			if (!afc->serverFormats[index].formatName)
+				return CHANNEL_RC_NO_MEMORY;
+		}
+	}
+
+	for (index = 0; index < afc->numServerFormats; index++)
+	{
+		format = &(afc->serverFormats[index]);
+
+		if (format->formatId == CF_UNICODETEXT)
+		{
+			if ((rc = ios_cliprdr_send_client_format_data_request(cliprdr, CF_UNICODETEXT)) !=
```

Comment:
Since you are only interested in text formats here, change just this to  `return ios_cliprdr_send_client_format_data_request(cliprdr, CF_UNICODETEXT)`.

Currently, you are only doing this, when that call fails. But if it does not (regular case), then you return `CHANNEL_RC_OK` in the end anyway. Since in this situation you already request the data, there is no point in further actions here.
In fact, it actually causes issues: If (and that is not unlikely) the server side advertises both `CF_UNICODETEXT` and `CF_TEXT` in the same format list, you will ask for data twice.
As format data requests are not coupled with ids, and since you set `afc->requestedFormatId` before sending the id, you may end up sending a request for `CF_UNICODETEXT` (where `afc->requestedFormatId` is set too), and sending a request for `CF_TEXT` (where `afc->requestedFormatId` is also set again).
Upon receiving the format data response, you obviously need `afc->requestedFormatId` again, but very likely, the server sent data for `CF_UNICODETEXT`, but for conversion, you use `CF_TEXT` here.

This is especially relevant for copying non-ASCII content (like non-english locale text). `CF_TEXT` is also dependent on the locale, `CF_UNICODETEXT` is not.

So, overall, in addition to the direct return, which I mentioned in this comment at the beginning. Do your search in two separate for loops.
In the first one, you look for `CF_UNICODETEXT`. In the second one, you look for `CF_TEXT`.
With the direct return, you will upon finding what you wanted, send the request, and just return with that.

Reply:
I changed the implementation according to your recommendations, thanks @pnowack! Please take a look.

Reply:
This wasn't addressed in your push (https://github.com/FreeRDP/FreeRDP/compare/5938d557f59c64fc49bcab1385a193f0aa011491..fc1bf8ff6d115a8a9783d87f17bc5160f5ec72bc). The situation is still the same here.